### DISCUSSION TITLE
Dotnet 生态图形框架列表

### DISCUSSION THREAD

**Discussion by Dynesshely** [BYJRK/DotNet-Discussions#13] [通用]
# Windows Only | 仅限 Windows 系统
- Winform
  `GitHub`: https://github.com/dotnet/winforms
  `免费` `开源` (.net core 及以后的版本, .net framework 时代的则不开源但免费)
  古早 Windows GUI 应用框架, 支持 .net framework 或是 .net core 以及后来的 .net 5 6 7
  界面布局使用 C# 描述, 通过 VS 设计器可以使用拖拽的方式快速设计原型, 快速产出结果
- WPF (Windows Presentation Fundation)
  `GitHub`: https://github.com/dotnet/wpf
  `免费` `开源` (.net core 及以后的版本, .net framework 时代的则不开源但免费)
  Windows 平台一种富有表现力的图形框架, 使用 XAML 语言描述界面
  底层使用 DirectX 渲染, 很容易即可制作炫酷界面
  推荐在实际开发中使用 MVVM, 且 WPF 框架对 MVVM 架构提供了良好支持
- WinUI (Windows UI Library)
  `GitHub`: https://github.com/microsoft/microsoft-ui-xaml
  `免费` `开源`
  微软新一代原生 Windows UI 框架，深度集成于 Windows App SDK
  支持 Win32 桌面应用和 UWP，采用 Fluent Design 设计语言
  底层使用 DirectX 渲染，提供现代化控件和交互体验

# Cross Platform | 跨平台图形框架
- [Avalonia](https://avaloniaui.net/)
  `GitHub`: https://github.com/avaloniaui/avalonia
  `免费` `开源`
  界面使用 XAML 语言描述, 支持 MVVM, 使用体验几乎等同于 WPF
  使用 Skia 自绘, 可以在不同平台呈现相同的界面效果
  支持 `Windows`, `Linux`, `MacOS`, `Android`, `iOS`, `Wasm (Web Assembly)`, `Apple Vision OS`, `Tizen OS`, 等平台
  Windows 支持 10 以上版本, 7 8 不受官方承诺, 但是大部分情况下可以运行
- [Uno](https://platform.uno/)
  `GitHub`: https://github.com/unoplatform/uno
  `免费` `开源`
  用于为 `Windows`, `iOS`, `Android`, `Wasm (Web Assembly)`, `MacOS` 和 `Linux` 构建单代码基应用程序的开源 UI 平台
  使用 XAML 语言描述界面, 同样是自绘界面, 所以可以在不同平台呈现相同的界面效果
  Gallery: https://gallery.platform.uno/
- [MAUI](https://dotnet.microsoft.com/zh-cn/apps/maui)
  `GitHub`: https://github.com/dotnet/maui
  `免费` `开源`
  支持 `Windows`, `MacOS`, `Android`, `iOS` 平台
  使用 XAML 语言描述界面, 但默认控件库将调用系统原生控件, 同样支持自绘控件来保证不同平台的界面效果
- [Silk.NET](https://dotnet.github.io/Silk.NET/)
  `GitHub`: https://github.com/dotnet/Silk.NET
  `免费` `开源`
  提供对低级图形 API 的封装, 包括 OpenGL, OpenCL, OpenAL, OpenXR, GLFW, SDL, Vulkan, Assimp, WebGPU, 以及 DirectX
  开发者可以轻易使用 C# 调用这些 API 来进行绘图
  可以产生较小的输出文件, 但是需要开发者自己绘制一切, 没有现成控件
- [OpenTK](https://opentk.net/)
  `GitHub`: https://github.com/opentk
  `免费` `开源`
  本质是 OpenGL 的 dotnet 封装, 使得开发者可以轻易使用 C# 调用 OpenGL 来进行绘图
  可以产生较小的输出文件, 但是需要开发者自己绘制一切, 不提供现成控件
- [Unity](https://unity.com/)
  `商业付费` `闭源`
  国内官网推荐访问: https://unity.cn/
  本质是一个游戏引擎, 而非传统 UI 框架
  开发者可以使用 Unity 制作非常炫酷的 3D 效果, 给用户沉浸的交互体验
  缺点是, 最终产出体积不同乐观
  使用 Mono 作为跨平台 C# 运行时, 目前听说计划转向 .NET 通用运行时
- [GoDot](https://godotengine.org/)
  `GitHub`: https://github.com/godotengine/godot
  `免费` `开源`
  也是一个游戏引擎, 支持使用 C# 作为交互脚本
  且能做到较小的打包体积
- [CPF](http://cpf.cskin.net/)
  `付费` `闭源` `商业源码授权`
  一个国产 .net 跨平台 UI 框架, 支持 `Windows`, `Linux`, `MacOS`, `Windows` 最低支持 Windows XP
  提供代码转换工具来将 Winform, WPF 的布局切换到 CPF
- [Qt](https://www.qt.io/)
  `商业模式尚不清楚` `开源程度尚不清楚`
  传统 C++ 跨平台 GUI 框架（现有 Python Binding），支持常见桌面平台与移动端，现已增加 Bridges for C#，尚在公测中：https://www.qt.io/blog/qt-bridges-public-beta-for-csharp
- [Jalium.UI](https://github.com/VeryJokerJal/Jalium.UI)
  `GitHub`: https://github.com/VeryJokerJal/Jalium.UI
  `免费` `开源` (MIT)
  GPU 加速的跨平台 UI 框架，面向 .NET 10
  提供类 WPF 的对象模型，使用 JALXAML 标记语言（支持 Razor 语法扩展）
  渲染后端：DirectX 12 (Windows)、Vulkan (Linux/Android)、Metal (macOS) 及软件回退
  包含 80+ 控件：图表、Ribbon、Docking、InkCanvas、WebView、Terminal、WindowsFormsHost 等
  支持 ClearType 子像素渲染、液态玻璃/丙烯酸/云母等视觉效果、GIF/APNG/WebP 动画
  内建独立扩展栈（Hosting/DI/Configuration/Logging/Metrics），无 Microsoft.Extensions 依赖

**Comment by SlimeNull** [comment]
Uno构建很难绷，不知道我这里是怎么回事，还原nuget包需要消耗太多时间了，搞得我完全没有耐心使用，也不知道现在有没有改进。

**Reply by Dynesshely** [reply]
如果能排除网络问题的话, 那么就是 uno 框架自身的问题了
如果你能提供更多信息的话, 我也许可以试着帮忙解决
不过我现在也没有深入使用 uno 框架, 且最近也没有听说 uno 官方相关的活动
感觉这个框架目前的状态比较不温不火

**Comment by Dynesshely** [comment]
更新日志 `2024.05.05`

- 添加了 Silk.NET 一节
- 迁移讨论到 `通用` 分类

**Reply by Dynesshely** [reply]
更新日志 `2026.05.23`
- 添加了 Qt `// 说实话，这个太意料外了，在一个没人在意的角落，Qt 悄悄宣布了对 C# 的支持。`

**Reply by Minokori** [reply]
感觉 Qt 应该是在押注 Linux 上.NET的界面框架市场, 毕竟 Linux 这一块感觉用 .NET 做桌面应用已经默认用 Avalonia 了. 还有 Windows Only 不推荐 WinUI 吗, 另外我还看到过一个 [JaliumUI](https://github.com/VeryJokerJal/Jalium.UI), 不知道有没有人实际体验过.

**Reply by Dynesshely** [reply]
> 感觉 Qt 应该是在押注 Linux 上.NET的界面框架市场, 毕竟 Linux 这一块感觉用 .NET 做桌面应用已经默认用 Avalonia 了. 还有 Windows Only 不推荐 WinUI 吗, 另外我还看到过一个 [JaliumUI](https://github.com/VeryJokerJal/Jalium.UI), 不知道有没有人实际体验过.

Qt 确实在图形界面上算是非常成熟了，来争取一下目前做桌面最常用的 dotnet 生态还是很合理的。

另外，感谢提醒，已经把 WinUI 和 JaliumUI 放进列表里面啦（不过我有在想，类似 JaliumUI 这类比较早期还没发展完善的 UI 框架加入进这个列表的话，列表还叫“C# 图形框架推荐”是不是不太合适了）。

**Reply by Dynesshely** [reply]
OK 哦，我把讨论贴改名成 “Dotnet 生态图形框架列表” 了。
